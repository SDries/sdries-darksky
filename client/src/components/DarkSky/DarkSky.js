import React, { Component } from 'react';
import DarkSkyAlerts from '../DarkSkyAlerts/DarkSkyAlerts';
import DarkSkyFlags from '../DarkSkyFlags/DarkSkyFlags';
import DarkSkyPoint from '../DarkSkyPoint/DarkSkyPoint';
import DarkSkyBlock from '../DarkSkyBlock/DarkSkyBlock';

class DarkSky extends Component {
  render() {
    if(!this.props.data) return null;

    let alerts, flags, currently, daily;
    if(this.props.data.alerts){
      alerts = <DarkSkyAlerts data={this.props.data.alerts} />;
    }
    if(this.props.data.flags){
      flags = <DarkSkyFlags data={this.props.data.flags} />;
    }
    if(this.props.data.currently){
      currently = <DarkSkyPoint point={this.props.data.currently} big />
    }
    if(this.props.data.daily){
      daily = <DarkSkyBlock block={this.props.data.daily} />
    }

    return (
      <div>
        {alerts}
        {currently}
        {daily}
        {flags}
      </div>
      );
  }
}

export default DarkSky;
