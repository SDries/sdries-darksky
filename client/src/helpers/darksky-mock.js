// https://darksky.net/dev/docs#response-alerts
export const alert = {
  description: 'A detailed description of the alert.',
  expires: 0,
  regions: ['region 1'],
  severity: "advisory",
  time: 0,
  title: 'A brief description of the alert.',
  url: 'https://www.google.com/',
};

export const point = {"time":1550764770,"summary":"Partly Cloudy","icon":"partly-cloudy-day","precipIntensity":0,"precipProbability":0,"temperature":9.57,"apparentTemperature":6.76,"dewPoint":6.79,"humidity":0.83,"pressure":1027.76,"windSpeed":20.95,"windGust":27.71,"windBearing":282,"cloudCover":0.4,"uvIndex":0,"visibility":8.08,"ozone":306.86};

// block with 2 days of data
export const block = {"summary":"No precipitation throughout the week, with high temperatures rising to 17Â°C on Wednesday.","icon":"clear-day","data":[{"time":1550703600,"summary":"Partly cloudy throughout the day.","icon":"partly-cloudy-day","sunriseTime":1550731823,"sunsetTime":1550769241,"moonPhase":0.57,"precipIntensity":0,"precipIntensityMax":0.0025,"precipIntensityMaxTime":1550757600,"precipProbability":0,"temperatureHigh":11.43,"temperatureHighTime":1550754000,"temperatureLow":5.1,"temperatureLowTime":1550815200,"apparentTemperatureHigh":11.43,"apparentTemperatureHighTime":1550754000,"apparentTemperatureLow":4.58,"apparentTemperatureLowTime":1550808000,"dewPoint":4.65,"humidity":0.86,"pressure":1026.81,"windSpeed":11.64,"windGust":32.17,"windGustTime":1550739600,"windBearing":244,"cloudCover":0.45,"uvIndex":2,"uvIndexTime":1550750400,"visibility":8.77,"ozone":314.75,"temperatureMin":2.68,"temperatureMinTime":1550725200,"temperatureMax":11.43,"temperatureMaxTime":1550754000,"apparentTemperatureMin":0.12,"apparentTemperatureMinTime":1550725200,"apparentTemperatureMax":11.43,"apparentTemperatureMaxTime":1550754000},{"time":1550790000,"summary":"Partly cloudy starting in the afternoon, continuing until evening.","icon":"partly-cloudy-day","sunriseTime":1550818102,"sunsetTime":1550855748,"moonPhase":0.61,"precipIntensity":0,"precipIntensityMax":0,"precipProbability":0,"temperatureHigh":13.84,"temperatureHighTime":1550847600,"temperatureLow":4.93,"temperatureLowTime":1550901600,"apparentTemperatureHigh":13.84,"apparentTemperatureHighTime":1550847600,"apparentTemperatureLow":2.07,"apparentTemperatureLowTime":1550901600,"dewPoint":6.98,"humidity":0.88,"pressure":1034.53,"windSpeed":3.3,"windGust":32.19,"windGustTime":1550872800,"windBearing":122,"cloudCover":0.22,"uvIndex":2,"uvIndexTime":1550833200,"visibility":16.09,"ozone":282.43,"temperatureMin":5.1,"temperatureMinTime":1550815200,"temperatureMax":13.84,"temperatureMaxTime":1550847600,"apparentTemperatureMin":4.58,"apparentTemperatureMinTime":1550808000,"apparentTemperatureMax":13.84,"apparentTemperatureMaxTime":1550847600}]};
