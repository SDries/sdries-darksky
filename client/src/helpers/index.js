import React from 'react';

export const formatUnixDate = (unixDate) => {
	var t = new Date(unixDate * 1000);
	return t.toLocaleDateString("en-US", { weekday: 'long', month: 'short', day: 'numeric' });
};

export const formatTemperature = (temperature) => {
	return Math.floor(temperature);
}

export const temperatureUnits = <span>&deg; C</span>;
