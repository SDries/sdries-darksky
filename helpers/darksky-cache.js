const cache = require('memory-cache');

const DarkskyCache = {
  getCachedResponse(longitude, latitude, time) {
    const key = this.getCacheKey(longitude, latitude, time);
    const cachedResponse = cache.get(key);
    if (cachedResponse) {
      return Promise.resolve(cachedResponse);
    }
    return Promise.resolve(null);
  },

  /**
   * creates a caching key for a longitude, latitude and day
   * @param  {string} longitude
   * @param  {string} latitude
   * @param  {string} day       YYYY-MM-DD string
   * @return {string}           caching key
   */
  getCacheKey(longitude, latitude, day) {
    return `${longitude}-${latitude}-${day}`;
  },


  /**
   * caches response for one day
   * @param  {string} longitude
   * @param  {string} latitude
   * @param  {string} dayString
   * @param  {object} response  response to be cached
   * @return {object}           returns the response
   */
  cacheTimemachineResponse(longitude, latitude, dayString, response) {
    const day = 24 * 60 * 60 * 1000;
    const key = this.getCacheKey(longitude, latitude, dayString);
    return cache.put(key, response, day);
  },
};

module.exports = DarkskyCache;
