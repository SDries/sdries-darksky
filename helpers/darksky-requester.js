const DarkSky = require('dark-sky');

module.exports = class DarkSkyRequester extends DarkSky {
  constructor(...args) {
    super(...args);

    this.darkSkyUnits = 'ca';
    this.darkSkyLanguage = 'en';
  }

  makeRequest(longitude, latitude, time = false) {
    const request = this.coordinates({ lat: latitude, lng: longitude });
    if (time) {
      request.time(time).exclude('minutely,hourly,currently,flags');
    } else {
      request.time(null).exclude('minutely,hourly');
    }
    return request
      .language(this.darkSkyLanguage)
      .units(this.darkSkyUnits)
      .get();
  }

  /**
   * API should return smaller, easier to handle, object
   * @param  {object} response Dark sky api response
   * @return {object}          Dark sky api response sub set
   */
  static filterApiResponse(response) {
    const allowedFields = ['currently', 'daily', 'alerts', 'flags'];

    return Object.keys(response)
      .filter(key => allowedFields.includes(key))
      .reduce((obj, key) => {
        obj[key] = response[key]; // eslint-disable-line no-param-reassign
        return obj;
      }, {});
  }
};
