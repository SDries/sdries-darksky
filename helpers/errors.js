module.exports = {
  formatErrorJson: { error: 'The request is not in a valid format.' },
  apiErrorJson: { error: 'Error during API call.' },
};
