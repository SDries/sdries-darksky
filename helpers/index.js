const Helpers = {
  /**
   * returns two digits (added a leading zero if needed)
   * @param  {number} number
   * @return {number}
   */
  formatNumberWithLeadingZero: number => (`0${number}`).slice(-2),

  /**
   * returns date formatted as YYYY-MM-DD
   * @param  {object} date date object
   * @return {string}      formatted string
   */
  getDayString: date => `${date.getFullYear()}-${Helpers.formatNumberWithLeadingZero(date.getMonth() + 1)}-${Helpers.formatNumberWithLeadingZero(date.getDate())}`,

  /**
   * returns array with date objects
   * @return {array} array of dates
   */
  getPast30Days: () => {
    const date = new Date();
    const days = [];

    for (let i = 1; i < 31; i += 1) {
      date.setDate(date.getDate() - 1);
      days.push(new Date(date));
    }
    return days;
  },
};

module.exports = Helpers;
