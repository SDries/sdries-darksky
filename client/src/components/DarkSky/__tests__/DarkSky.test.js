import React from 'react';
import DarkSky from '../DarkSky';
import ShallowRenderer from 'react-test-renderer/shallow';
import { point, block } from '../../../helpers/darksky-mock';

describe('react-darksky', () => {
  test('renders without data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<DarkSky />);
    const result = renderer.getRenderOutput();
    expect(result).toBeNull();
  });

  it('shallowly renders children correctly', () => {
    const apiMock = {
      alerts: [],
      flags: true,
      currently: point,
      daily: block,
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DarkSky data={apiMock} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
