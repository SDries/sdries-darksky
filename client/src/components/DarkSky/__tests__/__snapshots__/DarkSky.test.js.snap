// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-darksky shallowly renders children correctly 1`] = `
<div>
  <DarkSkyAlerts
    data={Array []}
  />
  <DarkSkyPoint
    big={true}
    point={
      Object {
        "apparentTemperature": 6.76,
        "cloudCover": 0.4,
        "dewPoint": 6.79,
        "humidity": 0.83,
        "icon": "partly-cloudy-day",
        "ozone": 306.86,
        "precipIntensity": 0,
        "precipProbability": 0,
        "pressure": 1027.76,
        "summary": "Partly Cloudy",
        "temperature": 9.57,
        "time": 1550764770,
        "uvIndex": 0,
        "visibility": 8.08,
        "windBearing": 282,
        "windGust": 27.71,
        "windSpeed": 20.95,
      }
    }
  />
  <DarkSkyBlock
    block={
      Object {
        "data": Array [
          Object {
            "apparentTemperatureHigh": 11.43,
            "apparentTemperatureHighTime": 1550754000,
            "apparentTemperatureLow": 4.58,
            "apparentTemperatureLowTime": 1550808000,
            "apparentTemperatureMax": 11.43,
            "apparentTemperatureMaxTime": 1550754000,
            "apparentTemperatureMin": 0.12,
            "apparentTemperatureMinTime": 1550725200,
            "cloudCover": 0.45,
            "dewPoint": 4.65,
            "humidity": 0.86,
            "icon": "partly-cloudy-day",
            "moonPhase": 0.57,
            "ozone": 314.75,
            "precipIntensity": 0,
            "precipIntensityMax": 0.0025,
            "precipIntensityMaxTime": 1550757600,
            "precipProbability": 0,
            "pressure": 1026.81,
            "summary": "Partly cloudy throughout the day.",
            "sunriseTime": 1550731823,
            "sunsetTime": 1550769241,
            "temperatureHigh": 11.43,
            "temperatureHighTime": 1550754000,
            "temperatureLow": 5.1,
            "temperatureLowTime": 1550815200,
            "temperatureMax": 11.43,
            "temperatureMaxTime": 1550754000,
            "temperatureMin": 2.68,
            "temperatureMinTime": 1550725200,
            "time": 1550703600,
            "uvIndex": 2,
            "uvIndexTime": 1550750400,
            "visibility": 8.77,
            "windBearing": 244,
            "windGust": 32.17,
            "windGustTime": 1550739600,
            "windSpeed": 11.64,
          },
          Object {
            "apparentTemperatureHigh": 13.84,
            "apparentTemperatureHighTime": 1550847600,
            "apparentTemperatureLow": 2.07,
            "apparentTemperatureLowTime": 1550901600,
            "apparentTemperatureMax": 13.84,
            "apparentTemperatureMaxTime": 1550847600,
            "apparentTemperatureMin": 4.58,
            "apparentTemperatureMinTime": 1550808000,
            "cloudCover": 0.22,
            "dewPoint": 6.98,
            "humidity": 0.88,
            "icon": "partly-cloudy-day",
            "moonPhase": 0.61,
            "ozone": 282.43,
            "precipIntensity": 0,
            "precipIntensityMax": 0,
            "precipProbability": 0,
            "pressure": 1034.53,
            "summary": "Partly cloudy starting in the afternoon, continuing until evening.",
            "sunriseTime": 1550818102,
            "sunsetTime": 1550855748,
            "temperatureHigh": 13.84,
            "temperatureHighTime": 1550847600,
            "temperatureLow": 4.93,
            "temperatureLowTime": 1550901600,
            "temperatureMax": 13.84,
            "temperatureMaxTime": 1550847600,
            "temperatureMin": 5.1,
            "temperatureMinTime": 1550815200,
            "time": 1550790000,
            "uvIndex": 2,
            "uvIndexTime": 1550833200,
            "visibility": 16.09,
            "windBearing": 122,
            "windGust": 32.19,
            "windGustTime": 1550872800,
            "windSpeed": 3.3,
          },
        ],
        "icon": "clear-day",
        "summary": "No precipitation throughout the week, with high temperatures rising to 17Â°C on Wednesday.",
      }
    }
  />
  <DarkSkyFlags
    data={true}
  />
</div>
`;
